Index: strace-5.9/tests/stack-fcall-0.c
===================================================================
--- strace-5.9.orig/tests/stack-fcall-0.c	2020-08-03 10:00:00.000000000 +0200
+++ strace-5.9/tests/stack-fcall-0.c	2020-10-06 11:12:07.163673799 +0200
@@ -7,7 +7,7 @@
 
 #include "stack-fcall.h"
 
-int
+int __attribute__ ((noinline))
 f0(int i, unsigned long f)
 {
 	f ^= (unsigned long) (void *) f0;
Index: strace-5.9/tests/stack-fcall-1.c
===================================================================
--- strace-5.9.orig/tests/stack-fcall-1.c	2020-08-03 10:00:00.000000000 +0200
+++ strace-5.9/tests/stack-fcall-1.c	2020-10-06 11:12:15.875605255 +0200
@@ -7,7 +7,7 @@
 
 #include "stack-fcall.h"
 
-int
+int __attribute__ ((noinline))
 f1(int i, unsigned long f)
 {
 	f ^= (unsigned long) (void *) f1;
Index: strace-5.9/tests/stack-fcall-2.c
===================================================================
--- strace-5.9.orig/tests/stack-fcall-2.c	2020-08-03 10:00:00.000000000 +0200
+++ strace-5.9/tests/stack-fcall-2.c	2020-10-06 11:12:22.700551557 +0200
@@ -7,7 +7,7 @@
 
 #include "stack-fcall.h"
 
-int
+int __attribute__ ((noinline))
 f2(int i, unsigned long f)
 {
 	f ^= (unsigned long) (void *) f2;
Index: strace-5.9/tests/stack-fcall-3.c
===================================================================
--- strace-5.9.orig/tests/stack-fcall-3.c	2020-08-03 10:00:00.000000000 +0200
+++ strace-5.9/tests/stack-fcall-3.c	2020-10-06 11:12:32.556474012 +0200
@@ -8,7 +8,7 @@
 #include <signal.h>
 #include "stack-fcall.h"
 
-int
+int __attribute__ ((noinline))
 f3(int i, unsigned long f)
 {
 	f ^= (unsigned long) (void *) f3;
